docs/developer/architecture/user_customization_and_persistence.rst

```diff
@@ -0,0 +1,224 @@
+following actions;
+
+  - Apply a salt state that reads the ConfigMap object, validates the schema
+    and checks the new values passed.
+
+  - Finally, we make sure that the updated values are picked up by their
+    respective consumers(this might require Pod restarts for changes to take
+    effect). The podSelector field has labels which can be used to detect which
+    deployments to restart.
+
+Note that, salt-states are used to sync data and update consumers
+of new configurations changes mainly because of the minimum effort it takes to
+setup this flow but the K8s Operator pattern could be use to replace
+configuration synchronization between user defined configurations and
+consumers.
+
+The Operator approach is much more complex, requires much more effort
+to realize and there is no real need for applying changes using this method
+because configuration changes are not frequent(for a typical MetalK8s admin,
+changing the number of replicas for a given deployment could happen once in 3
+months) as such, having an operator watch for object changes is not significant
+and not very useful at this point in time.
+
+**Using Operator architecture(Custom Controllers)**
+
+When using an Operator(a Custom Controller that works with CRDs), we create a
+Custom Resource Definition (CRD) that references a ConfigMap. Once a ConfigMap
+is updated by the user, then the Operator is designed to perform the following
+actions;
+
+  - The Operator is connected to the API server to watch for changes in the
+    ConfigMap.
+  - If the Operator detects a modified ConfigMap, it then determines which line
+    of action it should take which are;
+
+      - Extract the ConfigMap name and object fields
+      - Extract the pods associated to this ConfigMap based on it's labels
+      - Read and validate the ConfigMap data and schema
+      - If the ConfigMap is valid, update the respective pods such that they
+        pick-up new configs.
+      - If the ConfigMap is invalid, log the error and perform no further
+        action.
+
+Iteration 1
+~~~~~~~~~~~
+
+- Define and deploy new ConfigMap stores that will hold user configurations
+  as listed in the requirements
+- Template and render Deployment and Pod manifests that will make use of
+  this user customizable configurations using pillar values
+- Document how to change user configurations using kubectl
+- Create and deploy persistent storage volumes for Grafana dashboards and
+  datasources
+- Document how to create these persistent volume for Grafana dashboards and
+  datasources
+- Document how to add Prometheus alerting rules using a ConfigMap. This
+  documentation should explain to users that the Prometheus object has a
+  ruleSelector field, which selects the ConfigMaps from which rules files are
+  loaded from. This means that to add new Prometheus alert rules, Users need to
+  create new ConfigMaps containing alert rules with labels matching what was
+  defined in the Prometheus ruleSelector field.
```

Comment:
Ah! Actually, this is true if you replace `ConfigMap` with `PrometheusRule`, IIUC. Or maybe not in our version of `prometheus-operator`...? Didn't check

Reply:
Not tested, so let's leave it at editing the original CRD.